core/src/main/java/io/substrait/relation/HasExtension.java

```diff
@@ -0,0 +1,17 @@
+package io.substrait.relation;
+
+import io.substrait.io.substrait.extension.AdvancedExtension;
+import java.util.Optional;
+
+/**
+ * Used to indicate the presence of an {@link AdvancedExtension} directly on a relation
+ *
+ * <p>This exists outside of {@link Rel} because there are relations that do not have advanced
+ * extensions. See {@link ExtensionLeaf}, {@link ExtensionSingle}, {@link ExtensionMulti}).
```

Comment:
Which relations don't support advanced extensions? Is it only `io.substrait.relation.Set`? Would it simplify things to build this into Rel instead and throw if attempts are made to access for Rels that don't support it?